### ISSUE TITLE
Create a MsBackendSql instance from an existing on-disk SQLite file is too slow

### ISSUE THREAD

**Issue by plantton** [rformassspectrometry/MsBackendSql#23]
Scenario: We have an on-disk SQLite file (`C.db`) which held billions of spectra, and the size is more than 300 GB.

To create a MsBackendSql instance from `C.db`, we have to:

```R
conn1 <- dbConnect(SQLite(), "C.db")
b1 <- backendInitialize(MsBackendSqlDb(conn1))
```
Ideally, `backendInitialize` will create the MsBackendSql instance immediately. But the current process will take around one hour to initialize the MsBackendSql instance. After checking the steps (function `conn1 <- dbConnect(SQLite(), "C.db")` can return a `DBIConnection` object immediately.), the expensive step is that we [queried and ordered](https://github.com/rformassspectrometry/MsBackendSql/blob/master/R/MsBackendSqlDb.R#L198-L201) the underlying huge `msdata` table in SQLite file.  

Solution:  
```{r}
conn1 <- dbConnect(SQLite(), "C.db")  
rec1 <- peakRAM(length1 <- dbGetQuery(conn1, "SELECT MAX(_pkey)  FROM msdata"))
rec1  
Elapsed_Time_sec  
1  0.001
```

We can also print the number of rows in `msdata`:  
```R
length1  
  MAX(_pkey)  
1   42592313  
```

**Comment by plantton** [comment]
I will replace the current [queried and ordered](https://github.com/rformassspectrometry/MsBackendSql/blob/master/R/MsBackendSqlDb.R#L198-L201) code with this solution, and check the compatibility in the package. Do you have any suggestions? @sgibb

**Comment by jorainer** [comment]
I am a little confused. The *correct* way to initialize the backend would be 

```r
conn1 <- dbConnect(SQLite(), "C.db")
b1 <- backendInitialize(MsBackendSqlDb(), dbcon = conn1)
```
i.e. you should not pass an argument directly to the `MsBackendSqlDb()` constructor. The backend should always only be initialized by passing additional arguments to `backendInitialize`. And with that you should end up with [this SQL call](https://github.com/rformassspectrometry/MsBackendSql/blob/master/R/MsBackendSqlDb.R#L217-L219). It surprises me that this call should take one hour - but maybe your data is indeed that large? Could you test the time it takes to just run `dbGetQuery(conn1, "select _pkey from msdata")`?

**Comment by plantton** [comment]
We had an enhancement of MsBackenSqlDb constructor #13 .   

I just simply dbConnect to the `C.db` file, and used the query `"select _pkey from msdata"`:

```R
rec1 <- peakRAM(conn1 <- dbConnect(SQLite(), "C.db"))
rec1

Elapsed_Time_sec  Total_RAM_Used_MiB  Peak_RAM_Used_MiB
    0.002                0                  0.2

rec3 <- peakRAM(qry1 <- dbGetQuery(conn1, "select _pkey from msdata"))  

Elapsed_Time_sec  Total_RAM_Used_MiB  Peak_RAM_Used_MiB
    4288.276                162.5                  362.5
```

**Comment by jorainer** [comment]
I somehow don't have a good feeling to speed up loading of the object by assuming that `"_pkey"` is incrementally from 1 to max _pkey. I would then rather suggest to have an optional variable to choose whether fast loading or slow loading should be performed.

On the other hand, with a 300GB large SQLite database - aren't we hitting the limits of SQLite? Does it still make sense to use an SQLite database for that or maybe better a MySQL or PostgreSQL database instead?

**Comment by jorainer** [comment]
Short update: I created a `data.frame` with 10^8 rows and 3 columns, column `"id"` containing an integer from 1 to 10^8 and inserted that table to a SQLite and a MySQL database.

A call of `select id from msdata` took 28 seconds on the SQLite and 5.8 seconds on the MySQL database. Extracting the full data (`select * from msdata`) took 70 seconds on the SQLite and 16 seconds on the MySQL database. To me that shows that for really large data sets we should move to MySQL or some other database system.

Note that the MySQL database system I'm running is by no means optimized. I guess a clever admin could tune the database considerably.

**Comment by sgibb** [comment]
BTW: Shouldn't we create indices for some tables/queries? (I don't know whether this is possible in all SQL flavours.)

**Comment by jorainer** [comment]
Yep, the column `"_pkey"` should definitely have an index on it. Also, were possible, we should use cached queries - but if I'm not wrong @plantton is already using them anyway. indices and cached queries are available for all SQL databases.

**Comment by plantton** [comment]
MsBackendSql utilizes `_pkey` as the Integer  _Primary Key_ for table `msdata`. If I'm not wrong, the integer primary key will be twice as faster as ordinary indices on SQLite tables.  

3. [Benefits Of WITHOUT ROWID Tables](https://sqlite.org/withoutrowid.html#:~:text=In%20an%20ordinary%20SQLite%20table,is%20a%20true%20PRIMARY%20KEY.)  

[Is an index needed for a primary key in SQLite?](https://stackoverflow.com/questions/3379292/is-an-index-needed-for-a-primary-key-in-sqlite)

**Comment by plantton** [comment]
> Short update: I created a `data.frame` with 10^8 rows and 3 columns, column `"id"` containing an integer from 1 to 10^8 and inserted that table to a SQLite and a MySQL database.
> 
> A call of `select id from msdata` took 28 seconds on the SQLite and 5.8 seconds on the MySQL database. Extracting the full data (`select * from msdata`) took 70 seconds on the SQLite and 16 seconds on the MySQL database. To me that shows that for really large data sets we should move to MySQL or some other database system.
> 
> Note that the MySQL database system I'm running is by no means optimized. I guess a clever admin could tune the database considerably.

I'm wondering if the file size on-disk also matters for queries like `SELECT COUNT(*) FROM madata`. Even with the same number of rows, counting on a small dbfile is obviously faster than a huge dbfile.

**Comment by jorainer** [comment]
Yes, I agree. DB file size will have an impact.

**Comment by plantton** [comment]
I had a benchmarking on _cached queries_ over writing functions (using _PRAGMA_ mode in SQLite), but didn't implement it in the constructor.

**Comment by plantton** [comment]
> Short update: I created a `data.frame` with 10^8 rows and 3 columns, column `"id"` containing an integer from 1 to 10^8 and inserted that table to a SQLite and a MySQL database.
> 
> A call of `select id from msdata` took 28 seconds on the SQLite and 5.8 seconds on the MySQL database. Extracting the full data (`select * from msdata`) took 70 seconds on the SQLite and 16 seconds on the MySQL database. To me that shows that for really large data sets we should move to MySQL or some other database system.
> 
> Note that the MySQL database system I'm running is by no means optimized. I guess a clever admin could tune the database considerably.

So the conclusion is that, we will keep the current code, in case if we integrate other SQL DBMS into `MsBackendSql` in future.

**Comment by jorainer** [comment]
Yes, I agree @plantton . Also, if someone works with such a large data set he/she will also be used to wait a little more than usual.
